packages/ckeditor5-html-support/src/datafilter.ts

```diff
@@ -758,6 +761,67 @@ export default class DataFilter extends Plugin {
 			schema.register( 'htmlEmptyElement', {
 				inheritAllFrom: '$inlineObject'
 			} );
+
+			// Helper function to check for HTML attributes and remove element if none exist.
+			const removeElementIfNoHtmlAttributes = ( element: Element | Item, writer: Writer ) => {
+				// Check if the element still has any html* attributes.
+				const hasHtmlAttributes = Array
+					.from( element.getAttributeKeys() )
+					.some( key => key.startsWith( 'html' ) );
+
+				// If no html* attributes remain, remove the element.
+				if ( !hasHtmlAttributes ) {
+					writer.remove( element );
+					return true;
+				}
+
+				return false;
+			};
+
+			// Register a post-fixer that removes htmlEmptyElement when its htmlXX attribute is removed.
+			// See: https://github.com/ckeditor/ckeditor5/issues/18089
+			editor.model.document.registerPostFixer( writer => {
+				const changes = editor.model.document.differ.getChanges();
+				let changed = false;
+
+				for ( const change of changes ) {
+					if ( change.type === 'remove' ) {
+						continue;
+					}
+
+					// Look for removal of html* attributes.
+					if ( change.type === 'attribute' && change.attributeNewValue === null ) {
+						// Find htmlEmptyElement instances in the range that lost their html attribute.
+						for ( const { item } of change.range.getWalker() ) {
+							if ( !item.is( 'element', 'htmlEmptyElement' ) ) {
+								continue;
+							}
+
+							if ( removeElementIfNoHtmlAttributes( item, writer ) ) {
```

Comment:
You should collect all element references and remove them after iterating all the changes, as positions will not update in the following diff items.